File: July2022/python/interleavingString.py
@Jiganesh 

class Solution:

    # TC : O(N+M)
    # SC : O(N) 
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        
        
        dp = {}
        
        def helper(i=0, j=0 , k=0):
            
            
            if i==len(s1) and j == len(s2) and k==len(s3):
                return True
            
            
            if (i, j) in dp : return dp[(i,j)]
            
            ans = False
            if i< len(s1) and k<len(s3) and s1[i]==s3[k]:
                ans = ans or helper(i+1, j , k+1)
                
            if j< len(s2) and k< len(s3) and s2[j]==s3[k]:
                ans = ans or helper(i, j+1 , k+1)
                
            dp[(i,j)] = ans
            return ans
        
        
        return  helper()

File: July2022/C++/interleavingString.cpp
class Solution {
    private:
    bool f(int i,int j,int l,vector<vector<int>> &dp,string 
           &s1,string &s2,string &s3)
    {
        if(i==s1.length() and j==s2.length() and l==s3.length()) return 1;
        if(i>s1.length() or j>s2.length()) return 0;
        
        if(dp[i][j] != -1) return dp[i][j];
        
        // If both strings char equals to s3
        if(s1[i]==s3[l] and s2[j]==s3[l])
            return dp[i][j] = (f(i+1,j,l+1,dp,s1,s2,s3) or
                              f(i,j+1,l+1,dp,s1,s2,s3));
        
        // If s1 char equals to s3
        else if(s1[i]==s3[l]) return dp[i][j] = f(i+1,j,l+1,dp,s1,s2,s3);
    
        // If s2 char equals to s3
        else if(s2[j]==s3[l]) return dp[i][j] = f(i,j+1,l+1,dp,s1,s2,s3);
        
        return dp[i][j] = 0;
    }
public:
    bool isInterleave(string s1, string s2, string s3) {
        
        int n = s1.length(), m = s2.length(), l = s3.length();
        
        if(n+m != l) return 0;
        vector<vector<int>> dp(n+1,vector<int> (m+1,-1));
        return f(0,0,0,dp,s1,s2,s3);
    }
};
